Facet derive on enum with repr creates a compilation error

Hi,

This is somewhat related to #378 that got fixed recently (thanks!)

With facet 0.25.1, deriving the Facet trait on an enum with a "typed" repr, because it seems to assume the discriminant is always i64.

For example,

```rust
macro_rules! test_macro_u16 {
    () => {
        242u16
    };
}

const CONST_VALUE_U16: u16 = 142;

#[repr(u16)]
#[derive(facet::Facet)]
enum TestEnum {
    Value1 = 42u16,
    Value2 = CONST_VALUE_U16,
    Value3 = test_macro_u16!(),
}
```` 

fails to compile with:

```
error[E0308]: mismatched types
   --> facet-enum-repr/examples/test-facet.rs:12:14
    |
10  | #[derive(facet::Facet)]
    |          ------------ arguments to this method are incorrect
11  | enum TestEnum {
12  |     Value1 = 42u16,
    |              ^^^^^ expected `i64`, found `u16`
    |
note: method defined here
   --> /var/home/max/.local/share/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/facet-core-0.25.1/src/types/ty/enum_.rs:137:18
    |
137 |     pub const fn discriminant(mut self, discriminant: i64) -> Self {
    |                  ^^^^^^^^^^^^
help: change the type of the numeric literal from `u16` to `i64`
    |
12  -     Value1 = 42u16,
12  +     Value1 = 42i64,
    |

error[E0308]: mismatched types
   --> facet-enum-repr/examples/test-facet.rs:13:14
    |
10  | #[derive(facet::Facet)]
    |          ------------ arguments to this method are incorrect
...
13  |     Value2 = CONST_VALUE_U16,
    |              ^^^^^^^^^^^^^^^ expected `i64`, found `u16`
    |
note: method defined here
   --> /var/home/max/.local/share/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/facet-core-0.25.1/src/types/ty/enum_.rs:137:18
    |
137 |     pub const fn discriminant(mut self, discriminant: i64) -> Self {
    |                  ^^^^^^^^^^^^

error[E0308]: mismatched types
   --> facet-enum-repr/examples/test-facet.rs:3:9
    |
3   |         242u16
    |         ^^^^^^ expected `i64`, found `u16`
...
10  | #[derive(facet::Facet)]
    |          ------------ arguments to this method are incorrect
...
14  |     Value3 = test_macro_u16!(),
    |              ----------------- in this macro invocation
    |
note: method defined here
   --> /var/home/max/.local/share/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/facet-core-0.25.1/src/types/ty/enum_.rs:137:18
    |
137 |     pub const fn discriminant(mut self, discriminant: i64) -> Self {
    |                  ^^^^^^^^^^^^
    = note: this error originates in the macro `test_macro_u16` (in Nightly builds, run with -Z macro-backtrace for more info)
help: change the type of the numeric literal from `u16` to `i64`
    |
3   -         242u16
3   +         242i64
    |

```

--- a/facet-derive-emit/src/process_enum.rs
+++ b/facet-derive-emit/src/process_enum.rs
@@ -225,7 +225,7 @@ pub(crate) fn process_enum(parsed: Enum) -> TokenStream {
                         exprs.push(quote! {
                             ::facet::Variant::builder()
                                 #variant_attrs_tokens
-                                .discriminant(#discriminant_ts)
+                                .discriminant(#discriminant_ts as i64)
                                 .data(::facet::StructType::builder().repr(::facet::Repr::c()).unit().build())
                                 #maybe_doc
                                 .build()
@@ -271,7 +271,7 @@ pub(crate) fn process_enum(parsed: Enum) -> TokenStream {
                             ]};
                             ::facet::Variant::builder()
                                 #variant_attrs_tokens
-                                .discriminant(#discriminant_ts)
+                                .discriminant(#discriminant_ts as i64)
                                 .data(::facet::StructType::builder().repr(::facet::Repr::c()).tuple().fields(fields).build())
                                 #maybe_doc
                                 .build()
@@ -327,7 +327,7 @@ pub(crate) fn process_enum(parsed: Enum) -> TokenStream {
                             ]};
                             ::facet::Variant::builder()
                                 #variant_attrs_tokens
-                                .discriminant(#discriminant_ts)
+                                .discriminant(#discriminant_ts as i64)
                                 .data(::facet::StructType::builder().repr(::facet::Repr::c()).struct_().fields(fields).build())
                                 #maybe_doc
                                 .build()
@@ -415,7 +415,7 @@ pub(crate) fn process_enum(parsed: Enum) -> TokenStream {
                         exprs.push(quote! {
                             ::facet::Variant::builder()
                                 #variant_attrs_tokens
-                                .discriminant(#discriminant_ts)
+                                .discriminant(#discriminant_ts as i64)
                                 .data(::facet::StructType::builder().repr(::facet::Repr::c()).unit().build())
                                 #maybe_doc
                                 .build()
@@ -470,7 +470,7 @@ pub(crate) fn process_enum(parsed: Enum) -> TokenStream {
                             ]};
                             ::facet::Variant::builder()
                                 #variant_attrs_tokens
-                                .discriminant(#discriminant_ts)
+                                .discriminant(#discriminant_ts as i64)
                                 .data(::facet::StructType::builder().repr(::facet::Repr::c()).tuple().fields(fields).build())
                                 #maybe_doc
                                 .build()
@@ -527,7 +527,7 @@ pub(crate) fn process_enum(parsed: Enum) -> TokenStream {
                             ]};
                             ::facet::Variant::builder()
                                 #variant_attrs_tokens
-                                .discriminant(#discriminant_ts)
+                                .discriminant(#discriminant_ts as i64)
                                 .data(::facet::StructType::builder().repr(::facet::Repr::c()).struct_().fields(fields).build())
                                 #maybe_doc
                                 .build()

--- a/facet/src/sample_generated_code.rs
+++ b/facet/src/sample_generated_code.rs
@@ -1399,7 +1399,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for KitchenSinkEnum {
             [
                 crate::Variant::builder()
                     .name("UnitVariant")
-                    .discriminant(0i64)
+                    .discriminant(0i64 as i64)
                     .data(
                         crate::StructType::builder()
                             .repr(crate::Repr::c())
@@ -1421,7 +1421,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for KitchenSinkEnum {
                     };
                     crate::Variant::builder()
                         .name("TupleVariantSimple")
-                        .discriminant(1i64)
+                        .discriminant(1i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())
@@ -1467,7 +1467,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for KitchenSinkEnum {
                     };
                     crate::Variant::builder()
                         .name("TupleVariantMulti")
-                        .discriminant(2i64)
+                        .discriminant(2i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())
@@ -1508,7 +1508,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for KitchenSinkEnum {
                     };
                     crate::Variant::builder()
                         .name("StructVariant")
-                        .discriminant(3i64)
+                        .discriminant(3i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())
@@ -1532,7 +1532,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for KitchenSinkEnum {
                     };
                     crate::Variant::builder()
                         .name("SensitiveTupleVariant")
-                        .discriminant(4i64)
+                        .discriminant(4i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())
@@ -1566,7 +1566,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for KitchenSinkEnum {
                     };
                     crate::Variant::builder()
                         .name("StructVariantWithSensitiveField")
-                        .discriminant(5i64)
+                        .discriminant(5i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())
@@ -1588,7 +1588,9 @@ unsafe impl<'__facet> crate::Facet<'__facet> for KitchenSinkEnum {
                                                                                     }).build()
                         }]
                     };
-                    crate::Variant::builder().name("ArbitraryVariant").attributes(&[crate::VariantAttribute::Arbitrary("arbitrary")]).discriminant(6i64).data(crate::StructType::builder().repr(crate::Repr::c()).tuple().fields(fields).build()).doc(&[" A variant marked as arbitrary, potentially skipped during processing."]).build()
+                    crate::Variant::builder().name("ArbitraryVariant").attributes(&[crate::VariantAttribute::Arbitrary("arbitrary")]).discriminant(6i64
+                                                                    as
+                                                                    i64).data(crate::StructType::builder().repr(crate::Repr::c()).tuple().fields(fields).build()).doc(&[" A variant marked as arbitrary, potentially skipped during processing."]).build()
                 },
                 {
                     let fields: &'static [crate::Field] = &const {
@@ -1603,7 +1605,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for KitchenSinkEnum {
                     };
                     crate::Variant::builder()
                         .name("NestedEnumVariant")
-                        .discriminant(7i64)
+                        .discriminant(7i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())
@@ -2048,7 +2050,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for SubEnum {
             [
                 crate::Variant::builder()
                     .name("OptionA")
-                    .discriminant(0i64)
+                    .discriminant(0i64 as i64)
                     .data(
                         crate::StructType::builder()
                             .repr(crate::Repr::c())
@@ -2070,7 +2072,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for SubEnum {
                     };
                     crate::Variant::builder()
                         .name("OptionB")
-                        .discriminant(1i64)
+                        .discriminant(1i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())
@@ -2094,7 +2096,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for SubEnum {
                     };
                     crate::Variant::builder()
                         .name("SensitiveOption")
-                        .discriminant(2i64)
+                        .discriminant(2i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())
@@ -2119,7 +2121,7 @@ unsafe impl<'__facet> crate::Facet<'__facet> for SubEnum {
                     crate::Variant::builder()
                         .name("ArbitraryOption")
                         .attributes(&[crate::VariantAttribute::Arbitrary("arbitrary")])
-                        .discriminant(3i64)
+                        .discriminant(3i64 as i64)
                         .data(
                             crate::StructType::builder()
                                 .repr(crate::Repr::c())

--- a/facet/tests/derive.rs
+++ b/facet/tests/derive.rs
@@ -565,3 +565,23 @@ fn plain_tuple() {
         _ => panic!("Expected tuple to be a UserType::Struct"),
     }
 }
+
+#[test]
+fn test_macro_u16() {
+    macro_rules! test_macro_u16 {
+        () => {
+            242u16
+        };
+    }
+
+    const CONST_VALUE_U16: u16 = 142;
+
+    #[repr(u16)]
+    #[derive(Facet)]
+    #[allow(dead_code)]
+    enum TestEnum {
+        Value1 = 42u16,
+        Value2 = CONST_VALUE_U16,
+        Value3 = test_macro_u16!(),
+    }
+}

